providers/frame/common.ts

```diff
@@ -0,0 +1,120 @@
+import { File } from "../../codegen/frame/mod.ts"
+import { ss58, testUser } from "../../crypto/mod.ts"
+import * as $ from "../../deps/scale.ts"
+
+const DEFAULT_TEST_USER_COUNT = 100_000
+const DEFAULT_TEST_USER_INITIAL_FUNDS = 1_000_000_000_000_000_000
+
+export async function createCustomChainSpec(
+  bin: string,
+  chain: string,
+  networkPrefix: number,
+): Promise<string> {
+  const buildSpecCmd = new Deno.Command(bin, {
+    args: [
+      "build-spec",
+      "--disable-default-bootnode",
+      "--chain",
+      chain,
+    ],
+  })
+  const chainSpec = JSON.parse(new TextDecoder().decode((await buildSpecCmd.output()).stdout))
+  const balances: any[] = chainSpec.genesis.runtime.balances.balances
+  for (let i = 0; i < DEFAULT_TEST_USER_COUNT; i++) {
+    balances.push([
+      ss58.encode(networkPrefix, testUser(i).publicKey),
+      DEFAULT_TEST_USER_INITIAL_FUNDS,
+    ])
+  }
+  const customChainSpecPath = await Deno.makeTempFile({
+    prefix: `custom-${chain}-chain-spec`,
+    suffix: ".json",
+  })
+  await Deno.writeTextFile(customChainSpecPath, JSON.stringify(chainSpec, undefined, 2))
+  const buildSpecRawCmd = new Deno.Command(bin, {
+    args: [
+      "build-spec",
+      "--disable-default-bootnode",
+      "--chain",
+      customChainSpecPath,
+      "--raw",
+    ],
+  })
+  const chainSpecRaw = JSON.parse(
+    new TextDecoder().decode((await buildSpecRawCmd.output()).stdout),
+  )
+  const customChainSpecRawPath = await Deno.makeTempFile({
+    prefix: `custom-${chain}-chain-spec-raw`,
+    suffix: ".json",
+  })
+  await Deno.writeTextFile(customChainSpecRawPath, JSON.stringify(chainSpecRaw, undefined, 2))
+  return customChainSpecRawPath
+}
+
+export async function chainFileWithUsers(
+  file: File,
+  url: string,
+): Promise<File> {
+  return new File(`
+    ${file.codeRaw}
+
+    type ArrayOfLength<
+      T,
+      L extends number,
+      A extends T[] = [],
+    > = number extends L ? T[]
+      : L extends A["length"] ? A
+      : ArrayOfLength<T, L, [...A, T]>
+
+    export async function users<N extends number>(count: N): Promise<ArrayOfLength<C.Sr25519, N>>
+    export async function users(count: number): Promise<C.Sr25519[]> {
```

Comment:
Let's extract this into a `userFactory` fn, exposed from within `test_pairs`. `userFactory` can accept the `url` as an argument and return the `users` fn. This way, we generate less code:

```ts
export const users = C.testUserFactory(url)
```